// node-csv/index.js
//Full credit for the parser goes to:
//  http://www.bennadel.com/blog/1504-Ask-Ben-Parsing-CSV-Strings-With-Javascript-Exec-Regular-Expression-Command.htm

var fs = require('fs'), Csv = exports.Csv;

exports.createParser = function (delimiter, quote, quote_escape)
{
    return new Csv(delimiter, quote, quote_escape);
}

exports.createWriter = function (delimiter, quote, quote_escape)
{
    delimiter = delimiter || ',';
    quote = quote || '"';
    quote_escape = quote_escape || '"';

    function write(list)
    {
        var csv = '', i;
        list.forEach(function (line)
        {
            for (i = 0, l = line.length; i < l; i++)
            {
                line[i] = line[i].replace(quote, quote_escape + quote);
            }
            csv += quote + line.join(quote + delimiter + quote) + quote + '\r\n';
        });
        return csv;
    }

    function writeFile(file, list, callback)
    {
        fs.writeFile(file, write(list), callback);
    }

    return {
        write: write,
        writeFile: writeFile
    };
}

Csv = function (delimiter, quote, quote_escape)
{
    this.delimiter = delimiter || ',';
    this.quote = quote || '"';
    this.quote_escape = quote_escape || '"';
}

Csv.prototype.parseFile = function (path, fn)
{
    var self = this;

    fs.readFile(path, function (err, data)
    {
        if (err) throw err;
        self.parse(data, fn);
    });
}

Csv.prototype.parse = function (data, fn)
{

    try
    {

        if (data.slice(-1) !== "\n")
        {
            data = data + "\n"
        }

        var d = this.delimiter,
            e = this.quote_escape,
            q = this.quote;

        var pattern = new RegExp(
            '(' + d + '|\\r?\\n|\\r|^)' +
            '(?:' + q + '([^' + q + ']*(?:' + e + q + '[^' + q + ']*)*)"|' +
            '([^' + q + d + '\\r\\n]*))'

            , 'gi');

        var csv = [[]];

        var matches = null, strMatchedValue, matchedDelimiter;

        if (e == '\\') e = '\\\\';

        while (matches = pattern.exec(data))
        {

            matchedDelimiter = matches[1];

            if (matchedDelimiter.length && matchedDelimiter !== d)
            {
                csv.push([]);
            }

            if (matches[2])
            {
                strMatchedValue = matches[2].replace(new RegExp(e + q, 'g'), q);
            } else
            {
                strMatchedValue = matches[3];
            }

            csv[csv.length - 1].push(strMatchedValue);

        }

        csv.pop();

        if (fn)
        {
            fn(null, csv);
        } else
        {
            return csv;
        }

    } catch (e)
    {
        if (fn)
        {
            fn(e, []);
        } else
        {
            throw e;
        }
    }
}

Csv.prototype.mapFile = function (path, fn)
{
    var self = this;

    this.parseFile(path, function (err, data)
    {
        self.map(data, fn);
    });
}

Csv.prototype.map = function (csv, fn)
{
    try
    {

        var columns = csv.shift(), //Get column names from the first row
            obj, obj_pool = [],
            i = row = 0,
            col_count = columns.length;

        for (row = 0, len = csv.length; row < len; row++)
        {
            if (csv[row].length != col_count)
            {
                throw Error('CSV column count mismatch on line ' + (row + 2));
            }
            obj = {}
            for (i = 0; i < col_count; i++)
            {
                obj[columns[i]] = csv[row][i];
            }
            obj_pool.push(obj);
        }

        fn(null, obj_pool);

    } catch (e)
    {
        fn(e, []);
    }
}
